llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -28916,6 +28925,39 @@ SDValue AArch64TargetLowering::LowerVECTOR_HISTOGRAM(SDValue Op,
   return Scatter;
 }
 
+SDValue
+AArch64TargetLowering::LowerPARTIAL_REDUCE_ADD(SDValue Op,
+                                               SelectionDAG &DAG) const {
+  SDLoc DL(Op);
+  SDValue Acc = Op.getOperand(0);
+  SDValue Input = Op.getOperand(1);
+
+  EVT AccVT = Acc.getValueType();
+  EVT InputVT = Input.getValueType();
+
+  unsigned Opcode = Op.getOpcode();
+
+  if (AccVT.getVectorElementCount() * 4 == InputVT.getVectorElementCount()) {
+    unsigned IndexAdd = 0;
+    // ISD::MUL may have already been lowered, meaning the operands would be in
+    // different positions.
+    if (Input.getOpcode() != ISD::MUL)
+      IndexAdd = 1;
+    auto A = Input.getOperand(IndexAdd);
```

Comment:
This looks unsafe. If the `ISD::MUL` is not there then you can't assume it's a dot product?

Reply:
In lowering, the `ISD::MUL` is lowered before the partial reduction node. So I've now made it so if the `Input` opcode is NOT `ISD::MUL` and `Input`'s number of elements is 4 times greater than `Acc`'s number of elements, then it is expanded. This now happens in the DAG-combine.